File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-8.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {

        if (root == nullptr) return new TreeNode(val);
        
        TreeNode* curr = root;
        auto insert = [&](auto self) -> void {
            if (curr == nullptr) return;

            if (curr->val < val) {
                if (curr->right == nullptr) {
                    curr->right = new TreeNode(val);
                    return;
                }
                curr = curr->right;
                self(self);
            } else {
                if (curr->val > val) {
                    if (curr->left == nullptr) {
                        curr->left = new TreeNode(val);
                        return;
                    }
                    curr = curr->left;
                    self(self);
                }
            }
        };
        insert(insert);
        return root;

        }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.best = 0

        def dfs(node: TreeNode) -> None:
            if not node:
                return
            
            cur_max = find_deepest(node.left) + find_deepest(node.right)
            self.best = max(self.best, cur_max)

            dfs(node.left)
            dfs(node.right)

        
        def find_deepest(node: TreeNode) -> int:
            if not node: return 0
            return 1 + max(find_deepest(node.left), find_deepest(node.right))
        
        dfs(root)

        return self.best